src/platform/test/rom-test-main.c

```diff
@@ -0,0 +1,217 @@
+
+	mCoreConfigSetDefaultValue(&core->config, "idleOptimization", "remove");
+
+#ifdef M_CORE_GBA
+	if (core->platform(core) == mPLATFORM_GBA) {
+		((struct GBA*) core->board)->hardCrash = false;
+
+		_exitSwiImmediate = romTestOpts.exitSwiImmediate;
+		_returnCodeRegister = romTestOpts.returnCodeRegister;
+
+		if (_exitSwiImmediate == 3) {
+			// Hook into SWI 3 (shutdown)
+			struct mCoreCallbacks callbacks = {0};
+			callbacks.context = core;
+			callbacks.shutdown = (void(*)(void*)) _romTestSwi3Callback;
+			core->addCoreCallbacks(core, &callbacks);
+		} else {
+			// Custom SWI hooks
+			_armSwi16 = ((struct GBA*) core->board)->cpu->irqh.swi16;
+			((struct GBA*) core->board)->cpu->irqh.swi16 = _romTestSwi16;
+			_armSwi32 = ((struct GBA*) core->board)->cpu->irqh.swi32;
+			((struct GBA*) core->board)->cpu->irqh.swi32 = _romTestSwi32;
+		}
+	}
+#endif
+
+	bool cleanExit = true;
+	if (!mCoreLoadFile(core, args.fname)) {
+		cleanExit = false;
+		goto loadError;
+	}
+	if (args.patch) {
+		core->loadPatch(core, VFileOpen(args.patch, O_RDONLY));
+	}
+
+	struct VFile* savestate = NULL;
+
+	if (args.savestate) {
+		savestate = VFileOpen(args.savestate, O_RDONLY);
+	}
+
+	core->reset(core);
+
+	struct mCheatDevice* device;
+	if (args.cheatsFile && (device = core->cheatDevice(core))) {
+		struct VFile* vf = VFileOpen(args.cheatsFile, O_RDONLY);
+		if (vf) {
+			mCheatDeviceClear(device);
+			mCheatParseFile(device, vf);
+			vf->close(vf);
+		}
+	}
+
+	if (savestate) {
+		mCoreLoadStateNamed(core, savestate, 0);
+		savestate->close(savestate);
+	}
+
+	do {
+		core->runLoop(core);
+	} while (!_dispatchExiting);
+
+	core->unloadROM(core);
+
+loadError:
+	freeArguments(&args);
+	mCoreConfigDeinit(&core->config);
+	core->deinit(core);
+
+	return cleanExit ? _exitCode : 1;
+}
+
+static void _romTestShutdown(int signal) {
+	UNUSED(signal);
+	_dispatchExiting = true;
+}
+
+#ifdef M_CORE_GBA
+static void _romTestSwi3Callback(struct mCore* core) {
+	_exitCode = ((struct GBA*) core->board)->cpu->regs.gprs[_returnCodeRegister];
+	_dispatchExiting = true;
+}
+
+static void _romTestSwi16(struct ARMCore* cpu, int immediate) {
+	if (immediate == _exitSwiImmediate) {
+		_exitCode = cpu->regs.gprs[_returnCodeRegister];
+		_dispatchExiting = true;
+		return;
+	}
+	_armSwi16(cpu, immediate);
+}
+
+static void _romTestSwi32(struct ARMCore* cpu, int immediate) {
+	if (immediate == _exitSwiImmediate) {
+		_exitCode = cpu->regs.gprs[_returnCodeRegister];
+		_dispatchExiting = true;
+		return;
+	}
+	_armSwi32(cpu, immediate);
+}
+#endif
+
+static bool _parseRomTestOpts(struct mSubParser* parser, int option, const char* arg) {
+	struct romTestOpts* opts = parser->opts;
+	errno = 0;
+	switch (option) {
+	case 'S':
+		return _parseSwi(arg, &opts->exitSwiImmediate);
+	case 'R':
+		return _parseNamedRegister(arg, &opts->returnCodeRegister);
+	default:
+		return false;
+	}
+}
+
+static bool _parseSwi(const char* regStr, int* oSwi) {
+	long swi = strtol(regStr, NULL, 0);
```

Comment:
It might be a good idea to make sure there isn't junk following the string, just for completeness sake. The errno alone is not sufficient given how strto* is specified.